ProcessLib/ThermoMechanicalPhaseField/ThermoMechanicalPhaseFieldFEM.h

```diff
@@ -0,0 +1,384 @@
+};
+
+template <typename ShapeFunction, typename IntegrationMethod,
+          int DisplacementDim>
+class ThermoMechanicalPhaseFieldLocalAssembler : public ThermoMechanicalPhaseFieldLocalAssemblerInterface
+{
+public:
+    using ShapeMatricesType =
+        ShapeMatrixPolicyType<ShapeFunction, DisplacementDim>;
+    // Types for displacement.
+    // (Higher order elements = ShapeFunction).
+    using ShapeMatrices = typename ShapeMatricesType::ShapeMatrices;
+    using BMatricesType = BMatrixPolicyType<ShapeFunction, DisplacementDim>;
+
+    using NodalForceVectorType = typename BMatricesType::NodalForceVectorType;
+
+    using GlobalDimVectorType = typename ShapeMatricesType::GlobalDimVectorType;
+
+    ThermoMechanicalPhaseFieldLocalAssembler(ThermoMechanicalPhaseFieldLocalAssembler const&) = delete;
+    ThermoMechanicalPhaseFieldLocalAssembler(ThermoMechanicalPhaseFieldLocalAssembler&&) = delete;
+
+    ThermoMechanicalPhaseFieldLocalAssembler(
+        MeshLib::Element const& e,
+        std::size_t const /*local_matrix_size*/,
+        bool const is_axially_symmetric,
+        unsigned const integration_order,
+        ThermoMechanicalPhaseFieldProcessData<DisplacementDim>& process_data,
+        int const mechanics_related_process_id,
+        int const phase_field_process_id,
+        int const heat_conduction_process_id)
+        : _process_data(process_data),
+          _integration_method(integration_order),
+          _element(e),
+          _is_axially_symmetric(is_axially_symmetric),
+          _mechanics_related_process_id(mechanics_related_process_id),
+          _phase_field_process_id(phase_field_process_id),
+          _heat_conduction_process_id(heat_conduction_process_id)
+    {
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        _ip_data.reserve(n_integration_points);
+        _secondary_data.N.resize(n_integration_points);
+
+        auto const shape_matrices =
+            initShapeMatrices<ShapeFunction, ShapeMatricesType,
+                              IntegrationMethod, DisplacementDim>(
+                e, is_axially_symmetric, _integration_method);
+
+        SpatialPosition x_position;
+        x_position.setElementID(_element.getID());
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            // displacement (subscript u)
+            _ip_data.emplace_back(*_process_data.material);
+            auto& ip_data = _ip_data[ip];
+            ip_data.integration_weight =
+                _integration_method.getWeightedPoint(ip).getWeight() *
+                shape_matrices[ip].integralMeasure * shape_matrices[ip].detJ;
+
+            static const int kelvin_vector_size =
+                MathLib::KelvinVector::KelvinVectorDimensions<
+                    DisplacementDim>::value;
+            ip_data.eps.setZero(kelvin_vector_size);
+            ip_data.eps_prev.setZero(kelvin_vector_size);
```

Comment:
Initialization of the `_prev` quantities is not necessary. Better to use `resize` like it is done in PhaseFieldProcess. Initialization of these quantities conseils improper "push_back_state" implementation or other initial conditions implementation.